File: packages/sdk-python/frameforge/scene.py
"""Scene builder — the top-level container for a FrameForge video."""

from __future__ import annotations

import json
import subprocess
import tempfile
from pathlib import Path
from typing import Any

from frameforge.codegen import generate_html


class Scene:
    """A video scene with elements and animations."""

    def __init__(
        self,
        *,
        width: int = 1920,
        height: int = 1080,
        fps: int = 30,
        duration: float,
        background: str = "#000000",
    ):
        self.width = width
        self.height = height
        self.fps = fps
        self.duration = duration
        self.background = background
        self.elements: list[Any] = []
        self.audio: list[dict] = []

    def add(self, *elements: Any) -> Scene:
        """Add elements to the scene."""
        for el in elements:
            self.elements.append(el)
        return self

    def add_audio(
        self, src: str, *, start_at: float = 0, volume: float = 1.0
    ) -> Scene:
        """Add an audio track to the scene."""
        self.audio.append({"src": src, "startAt": start_at, "volume": volume})
        return self

    def to_manifest(self, output: str = "./output.mp4") -> dict:
        """Generate a scene manifest dict."""
        return {
            "version": "1.0",
            "canvas": {
                "width": self.width,
                "height": self.height,
                "fps": self.fps,
                "duration": self.duration,
                "background": self.background,
            },
            "entry": "./scene.html",
            "audio": self.audio,
            "render": {
                "codec": "h264",
                "quality": "high",
                "pixelFormat": "yuv420p",
                "output": output,
            },
        }

    def render(self, output: str = "./output.mp4") -> str:
        """Generate HTML and manifest, then invoke the FrameForge CLI."""
        with tempfile.TemporaryDirectory(prefix="frameforge-") as tmpdir:
            tmpdir_path = Path(tmpdir)

            # Generate HTML
            html = generate_html(self)
            html_path = tmpdir_path / "scene.html"
            html_path.write_text(html, encoding="utf-8")

            # Generate manifest
            manifest = self.to_manifest(str(Path(output).resolve()))
            manifest["entry"] = str(html_path)
            manifest_path = tmpdir_path / "scene.json"
            manifest_path.write_text(
                json.dumps(manifest, indent=2), encoding="utf-8"
            )

            # Call frameforge CLI
            result = subprocess.run(
                ["npx", "frameforge", "render", str(manifest_path)],
                capture_output=True,
                text=True,
            )

            if result.returncode != 0:
                raise RuntimeError(
                    f"FrameForge render failed:\n{result.stderr}"
                )

            return str(Path(output).resolve())

File: packages/sdk-ts/src/scene.ts
import type { SceneManifest } from "@frameforge/core";
import { generateHTML } from "./codegen.js";
import { writeFile, mkdir } from "node:fs/promises";
import { resolve, dirname } from "node:path";

export interface SceneOptions {
  width?: number;
  height?: number;
  fps?: number;
  duration: number;
  background?: string;
}

export interface SceneElement {
  type: string;
  props: Record<string, any>;
  animations: Animation[];
}

export interface Animation {
  property: string;
  keyframes: Record<number, number | string>;
  easing?: (t: number) => number;
  easingName?: string;
}

export class Scene {
  readonly width: number;
  readonly height: number;
  readonly fps: number;
  readonly duration: number;
  readonly background: string;
  readonly elements: SceneElement[] = [];

  constructor(options: SceneOptions) {
    this.width = options.width ?? 1920;
    this.height = options.height ?? 1080;
    this.fps = options.fps ?? 30;
    this.duration = options.duration;
    this.background = options.background ?? "#000000";
  }

  add(...elements: Array<{ toSceneElement(): SceneElement }>) {
    for (const el of elements) {
      this.elements.push(el.toSceneElement());
    }
    return this;
  }

  toManifest(outputPath: string = "./output.mp4"): SceneManifest {
    return {
      version: "1.0",
      canvas: {
        width: this.width,
        height: this.height,
        fps: this.fps,
        duration: this.duration,
        background: this.background,
      },
      entry: "./scene.html",
      audio: [],
      render: {
        codec: "h264",
        quality: "high",
        pixelFormat: "yuv420p",
        output: outputPath,
      },
    };
  }

  /**
   * Generate HTML and scene manifest files, then invoke the renderer.
   */
  async render(outputPath: string = "./output.mp4"): Promise<string> {
    const html = generateHTML(this);
    const manifest = this.toManifest(outputPath);

    // Write generated files to a temp directory
    const outDir = resolve(dirname(outputPath), ".frameforge-tmp");
    await mkdir(outDir, { recursive: true });

    const htmlPath = resolve(outDir, "scene.html");
    const manifestPath = resolve(outDir, "scene.json");

    await writeFile(htmlPath, html, "utf-8");
    manifest.entry = htmlPath;
    await writeFile(manifestPath, JSON.stringify(manifest, null, 2), "utf-8");

    // Dynamic import to avoid circular dependency at build time
    const { render: coreRender } = await import("@frameforge/core");
    return coreRender({ input: manifestPath, output: outputPath });
  }
}
